### PR TITLE
vignette on formats precedence

### LINKED ISSUE #366: Add vignette on formats precedence and NA handling
This will probably also involve documentation additions in formatters in terms of the underlying behavior

### PR DESCRIPTION
**Pull request by Melkiades**
Fixes #366 
@wwojciech fyi

### CODE REVIEW

File: vignettes/formats_precedence.Rmd [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+The examples shown in this vignette are based on the example `ADSL` dataset, i.e. a demographic table that summarizes the variables content for different population subsets (encoded in the columns).
+
+```{r}
+ADSL <- ex_adsl
+```
+
+Note that all `ex_*` data which is currently attached to the `rtables` package is provided by the [`formatters`](https://insightsengineering.github.io/formatters/) package and was created using the publicly available [`random.cdisc.data`](https://insightsengineering.github.io/random.cdisc.data/) R package.
+
+### Basic formats precedence and inheritance rules.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:43]:
```suggestion
### Basic formats precedence and inheritance rules
```

File: vignettes/formats_precedence.Rmd [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+The format in which numbers are printed can be specified by the user in a few different places. In the context of the precedence, it is important at which level of the split hierarchy formats are specified. In general, there are two such levels: the **cell** level or so called **parent table** level. The concept of parent table results from the way in which `rtables` package stores resulting tables. It models the resulting tables as hierarchical, tree-like objects which support sibling sub-tables, arbitrary splitting or grouping of data in row and column dimensions and cells containing multiple values. Particularly noteworthy in this context is the fact that the actual table splitting occurs in a row-dominant way (even if the column split is present in the layout). `rtables` provides user-end function `table_structure()` that prints the structure of a given table object. For a quick illustration, consider the following example.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:45]: a bit verbose

File: vignettes/formats_precedence.Rmd [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  analyze(vars = "AGE", afun = mean)
+
+adsl_analyzed <- build_table(lyt, ADSL)
+adsl_analyzed
+table_structure(adsl_analyzed)
+```
+
+For this reporting table, there are 4 sub-tables under `SEX` table, i.e. `F`, `M`, `U`, and `UNDIFFERENTIATED`. Each of these sub-tables has one sub-table `AGE`. So, e.g. for the top sub-table `AGE`, its nearest parent table is `F`.
+
+The concept of formats precedence and inheritance adopted by the `rtables` is quote straightforward. The cell level is the most specific level at which the format can be specified be the user. If the cell format is not explicitly specified, it is then inherited from its nearest parent table for which the format was chosen by the user (if any). If no any specific format has been selected by the user, then the default format will be applied. The default format is "xx" and it yields the same formatting as `as.character()` function. In summary, the precedence-inheritance chain looks like below.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:60]:
```suggestion
The concept of formats precedence and inheritance in `rtables` can have multiple levels of parent tables and a final cell level. The cell level is the most specific and lowest level at which the format can be specified by the user. If the cell format is not explicitly specified, it is then inherited from its nearest parent table for which the format was chosen by the user (if any). If no specific format has been inserted by the user, then the default format will be applied. The default format is "xx" and it yields the same formatting as `as.character()` function. In summary, the precedence-inheritance chain looks like below.
```

File: vignettes/formats_precedence.Rmd [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+```{r}
+lyt0 <- basic_table() %>%
+  split_cols_by("ARM") %>%
+  analyze(vars = "AGE", afun = mean)
+
+build_table(lyt0, ADSL)
+```
+
+### Cell format.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:78]:
```suggestion
### Cell format
```

File: vignettes/formats_precedence.Rmd [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+For this reporting table, there are 4 sub-tables under `SEX` table, i.e. `F`, `M`, `U`, and `UNDIFFERENTIATED`. Each of these sub-tables has one sub-table `AGE`. So, e.g. for the top sub-table `AGE`, its nearest parent table is `F`.
+
+The concept of formats precedence and inheritance adopted by the `rtables` is quote straightforward. The cell level is the most specific level at which the format can be specified be the user. If the cell format is not explicitly specified, it is then inherited from its nearest parent table for which the format was chosen by the user (if any). If no any specific format has been selected by the user, then the default format will be applied. The default format is "xx" and it yields the same formatting as `as.character()` function. In summary, the precedence-inheritance chain looks like below.
+```
+cell <- parent_table <- parent_table <- ... <- parent_table
+```
+In case when the format is specified by the user in more than one place, the one which is closer to the `cell` in  the precedence-inheritance chain will be applied. In the following sections of this vignette, we illustrate the formats precedence rules with a few examples.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:64]:
```suggestion
In case when the format is specified by the user in more than one place, the one which is closer to the `cell` in the precedence-inheritance chain will be applied. In the following sections of this vignette, we illustrate the format precedence rules with a few examples.
```

File: vignettes/formats_precedence.Rmd (hunk 2 above)
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:45]:
```suggestion
The format in which numbers are printed can be specified by the user in a few different places. In the context of the precedence, it is important at which level of the split hierarchy formats are specified. In general, there are two such levels: the **cell** level or the so-called **parent table** level. The concept of parent table results from the way in which `rtables` package stores resulting tables. It models the resulting tables as hierarchical, tree-like objects which support sibling sub-tables, arbitrary splitting or grouping of data in row and column dimensions, and cells containing multiple values. Particularly noteworthy in this context is the fact that the actual table splitting occurs in a row-dominant way (even if the column split is present in the layout). `rtables` provides user-end function `table_structure()` that prints the structure of a given table object. For a quick illustration, consider the following example.
```

File: vignettes/formats_precedence.Rmd [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+```{css, echo=FALSE}
+.reveal .r code {
+  white-space: pre;
+}
+```
+
+## Formats Precedence
+
+A user of `rtables` package can specify the format in which the numbers in the reporting tables are printed. Formatting functionally is provided by the [`formatters`](https://insightsengineering.github.io/formatters/) R package and available formats can be listed with `formatters::list_valid_format_labels()` function.
+The format can be specified by the user in a few different places. It may happen that for a single table layout, the format is specified in more than one place. In such a case, the final format that will be applied, depends on formats precedence rules defined by `rtables`. In this vignette, we describe basic rules of `rtables` formats precedence.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:27]:
```suggestion
The format can be specified by the user in a few different places. It may happen that for a single table layout, the format is specified in more than one place. In such a case, the final format that will be applied depends on formats precedence rules defined by `rtables`. In this vignette, we describe the basic rules of `rtables` formats precedence.
```

File: vignettes/formats_precedence.Rmd [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    in_rows(
+      "Mean" = rcell(mean(x)),
+      .formats = "xx.xx"
+    )
+  })
+
+build_table(lyt1a, ADSL)
+```
+
+If the format is specified in both of these places at the same time, the one specified through `in_rows()`, has a higher precedence. Technically, in this case, the format defined in `rcell()` will simply be overwritten by the one defined in `in_rows()`.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:103]:
```suggestion
If the format is specified in both of these places at the same time, the one specified through `in_rows()`, has higher precedence. Technically, in this case, the format defined in `rcell()` will simply be overwritten by the one defined in `in_rows()`.
```

File: vignettes/formats_precedence.Rmd [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+    in_rows(
+      "Mean" = rcell(mean(x), format = "xx.xxx"),
+      .formats = "xx.xx"
+    )
+  })
+
+build_table(lyt2, ADSL)
+```
+
+### Parent table format and inheritance.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:118]:
```suggestion
### Parent table format and inheritance
```

File: vignettes/formats_precedence.Rmd [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+    },
+    format = "xx.x"
+  )
+
+build_table(lyt5, ADSL)
+```
+
+## `NA` Handling
+
+Consider the following layout and the resulting table created.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:180]:
```suggestion
Consider the following layout and the resulting table created:
```

File: vignettes/formats_precedence.Rmd [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+```{r}
+lyt6 <- basic_table() %>%
+  split_cols_by("ARM") %>%
+  split_rows_by("SEX") %>%
+  analyze(vars = "AGE", afun = mean, format = "xx.xx")
+
+build_table(lyt6, ADSL)
+```
+
+In the output, the cell corresponding to `UNDIFFERENTIATED` level of `SEX` and `B: Placebo drug` level of `ARM` categorical variables is displayed as `NA`. This happened because the cell's value was `NA`. `rtables` allows user to specify a string which should be displayed when formatted if the cell's value(s) are all `NA`. Similar to the configuration of formatting for numbers, user can specify the `NA` string in a few different places, at the **cell** level or **parent table** level. The `NA` string precedence and inheritance rules are the same as those for number formats precedence, described in the previous section of this vignette. We will illustrate them with a few examples.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:191]:
```suggestion
In the output, the cell corresponding to `UNDIFFERENTIATED` level of `SEX` and `B: Placebo drug` level of `ARM` categorical variables is displayed as `NA`. This happened because there were no values for the mean to calculate that respected these two conditions. `rtables` allows the user to specify a string to display when cell values are `NA`. Similar to formats for numbers, the user can specify a string to replace `NA` with the parameter `format_na_str` or `.format_na_str`. This can be specified at the **cell** or **parent table** level. The `NA` string precedence and inheritance rules are the same as those for number formats precedence, described in the previous section of this vignette. We will illustrate them with a few examples.
```

File: vignettes/formats_precedence.Rmd [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+### Cell `NA` string.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:193]:
```suggestion
### Cell `NA` string
```

File: vignettes/formats_precedence.Rmd [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+
+At the cell level, the specification of the `NA` string is through the `format_na_str` argument of `rcell()` function or `.format_na_str` argument of `in_rows()` function.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:195]:
```suggestion
At the cell level, it is possible to replace `NA` values with a custom string by means of the `format_na_str` parameter in `rcell()` or `.format_na_str` parameter in `in_rows()`.
```

File: vignettes/formats_precedence.Rmd [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+    in_rows(
+      "Mean" = rcell(mean(x), format = "xx.xx"),
+      .format_na_strs = "<MISSING>"
+    )
+  })
+
+build_table(lyt7a, ADSL)
+```
+
+If the `NA` string is specified in both of these places at the same time, the one specified through `in_rows()`, has a higher precedence. Technically, in this case, the `NA` string defined in `rcell()` will simply be overwritten by the one defined in `in_rows()`.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:220]:
```suggestion
If the `NA` string is specified in both of these places at the same time, the one specified with `in_rows()`, has higher precedence. Technically, in this case, the `NA` replacement string defined in `rcell()` will simply be overwritten by the one defined in `in_rows()`.
```

File: vignettes/formats_precedence.Rmd [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+    in_rows(
+      "Mean" = rcell(mean(x), format = "xx.xx", format_na_str = "<missing>"),
+      .format_na_strs = "<MISSING>"
+    )
+  })
+
+build_table(lyt8, ADSL)
+```
+
+### Parent table `NA` string and inheritance.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:236]:
```suggestion
### Parent table replacement of `NA` values and inheritance principles
```

File: vignettes/formats_precedence.Rmd [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+In addition to the cell level, the `NA` string can be specified at the parent table level(s). Then, if no `NA` string has been specified by the user for a cell, the string from the nearest parent split table will be inherited (if any).
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:238]:
```suggestion
In addition to the cell level, the string replacement for `NA` values can be specified at the parent table level(s). Then, if no `NA` string has been specified by the user for a cell, the string from the nearest parent split table will be inherited (if any).
```

File: vignettes/formats_precedence.Rmd [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+```{r}
+lyt9 <- basic_table() %>%
+  split_cols_by("ARM") %>%
+  split_rows_by("SEX") %>%
+  analyze(vars = "AGE", mean, format = "xx.xx", na_str = "not available")
+
+build_table(lyt9, ADSL)
+```
+
+If the `NA` string was specified also at the cell level, then the one set at the parent table level is ignored for this cell, as the cell level settings are treated as the most specific in this case.
```
Comment (Melkiades) [inline vignettes/formats_precedence.Rmd:249]:
```suggestion
If the replacement string for `NA` values was specified also at the cell level, then the one set at the parent table level is ignored for this cell, as the cell level settings are treated as the most specific in this case.
```

**Review by Melkiades — COMMENTED** [review]
I think it is very good @wwojciech!! It is quite detailed and clear. I substituted in general `NA` string with a more precise description of it and changed some titles and phrases here and there. I would only suggest simplifying the introductory paragraph which can be a bit overwhelmingly complicated for this vignette. I would try to describe in simpler terms the table structure

File: vignettes/formats_precedence.Rmd [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+
+The concept of formats precedence and inheritance in `rtables` can have multiple levels of parent tables and a final cell level. The cell level is the most specific and lowest level at which the format can be specified by the user. If the cell format is not explicitly specified, it is then inherited from its nearest parent table for which the format was chosen by the user (if any). If no specific format has been selected by the user, then the default format will be applied. The default format is "xx" and it yields the same formatting as `as.character()` function. In summary, the precedence-inheritance chain looks like below.
+```
+cell <- parent_table <- parent_table <- ... <- parent_table
+```
+In case when the format is specified by the user in more than one place, the one which is closer to the `cell` in the precedence-inheritance chain will be applied. In the following sections of this vignette, we illustrate the format precedence rules with a few examples.
```
Comment (gmbecker) [inline vignettes/formats_precedence.Rmd:64]: Please use "most specific" rather than "closest to cell" when describing which will be apply to a particular value.

File: vignettes/formats_precedence.Rmd [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+    in_rows(
+      "Mean" = rcell(mean(x)),
+      .formats = "xx.xx"
+    )
+  })
+
+build_table(lyt1a, ADSL)
+```
+
+If the format is specified in both of these places at the same time, the one specified through `in_rows()`, has higher precedence. Technically, in this case, the format defined in `rcell()` will simply be overwritten by the one defined in `in_rows()`.
```
Comment (gmbecker) [inline vignettes/formats_precedence.Rmd:103]: Please explicitly say that this is because the format (and na_str) specified in in_rows is applied to the *cells* not the rows (overriding the previously specified cell specific values), which means that the precedence rules described above are still in place.

File: vignettes/formats_precedence.Rmd [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+  split_cols_by("ARM") %>%
+  analyze(vars = "AGE", mean, format = "xx.x")
+
+build_table(lyt3, ADSL)
+```
+
+If the cell format was specified too, then the parent table format is ignored for this cell, since the cell format is treated as the most specific.
+
+```{r}
+lyt4 <- basic_table() %>%
```
Comment (gmbecker) [inline vignettes/formats_precedence.Rmd:133]: Probably don't need this example, just the one below it that shows a cell that does inherit and a cell that doesn't  coming from the same analyze block

**Review by gmbecker — CHANGES_REQUESTED** [review]
I think we will probably want to come back and rework this (format and na_str have identical behavior, so we probably don't need them split out into separate sections), but for now, just make the requested changes and we will put this in the upcoming release, as it is correct and helpful.

File: vignettes/formats_precedence.Rmd (hunk 19 above)
Comment (wwojciech) [reply vignettes/formats_precedence.Rmd:133]: Thanks @gmbecker 
Hmm, to be honest I find layouts 4 and 4a valuable to this vignette, as they are somehow in between layouts 3 and 5. So I would prefer to keep it. However, if you sure that they should be removed - let me know and I will do it. Tx!

Comment (gmbecker) [reply vignettes/formats_precedence.Rmd:133]: 4 doesn't convey any information that 4a doesn't also communicate equally well. We don't want documentation to be longer than it needs to be because there's already a lot of it and there will be more in the future. @wwojciech

Comment (wwojciech) [reply vignettes/formats_precedence.Rmd:133]: I will remove it then. Tx for additional argumentation.

**Review by gmbecker — APPROVED** [review]
I will do some massaging after merge but overall this looks good enough to put into the release.
